src/daemon/daemon.cpp

```diff
@@ -1964,7 +2006,27 @@ void mp::Daemon::create_vm(const CreateRequest* request, grpc::ServerWriter<Crea
                         break;
                     }
                 }
-                auto vm_desc = to_machine_desc(request, name, checked_args.mem_size, checked_args.disk_space, mac_addr,
+
+                mp::MemorySize image_size = get_image_size(vm_image);
+                mp::MemorySize disk_space{};
+
+                if (checked_args.disk_space.in_bytes() == 0)
+                {
+                    auto default_disk_size_as_struct = mp::MemorySize(mp::default_disk_size);
+                    disk_space = image_size < default_disk_size_as_struct ? default_disk_size_as_struct : image_size;
+                }
+                else if (checked_args.disk_space < image_size)
+                {
+                    throw std::runtime_error(fmt::format("requested image size ({} bytes) is smaller than image "
+                                                         "size ({} bytes)",
+                                                         checked_args.disk_space.in_bytes(), image_size.in_bytes()));
```

Comment:
Could this be done earlier? With how it is now, we copy the image just to then throw that the requested size was too small.

Reply:
As discussed, doing it earlier would imply a major refactor.